social_core/backends/id4me.py

```diff
@@ -0,0 +1,270 @@
+        })
+        if self.STATE_PARAMETER and state:
+            params['state'] = state
+        if self.RESPONSE_TYPE:
+            params['response_type'] = self.RESPONSE_TYPE
+        return params
+
+    def auth_url(self):
+        if not self.data.get('identity', ''):
+            raise AuthMissingParameter(self, 'identity')
+        identity = self.data.get('identity')
+        if not is_valid_domain(identity):
+            raise AuthForbidden(self)
+        openid_configuration = self.get_identity_record(identity)
+        if 'iss' not in openid_configuration:
+            raise AuthUnreachableProvider(self)
+        self.strategy.session_set(self.name + '_authority', openid_configuration['iss'])
+        self.strategy.session_set(self.name + '_identity', identity)
+        return super(ID4meBackend, self).auth_url()
+
+    def auth_complete_params(self, state=None):
+        data = {
+            'grant_type': 'authorization_code',
+            'code': self.data.get('code', ''),
+            'redirect_uri': self.get_redirect_uri()
+        }
+        return '&'.join(["{}={}".format(key, value) for key, value in data.items()])
+
+    def auth_complete_credentials(self):
+        return self.get_key_and_secret()
+
+    def validate_claims(self, id_token):
+        utc_timestamp = timegm(datetime.datetime.utcnow().utctimetuple())
+
+        if 'nbf' in id_token and utc_timestamp < id_token['nbf']:
+            raise AuthTokenError(self, 'Incorrect id_token: nbf')
+
+        # Verify the token was issued in the last 10 minutes
+        iat_leeway = self.setting('ID_TOKEN_MAX_AGE', self.ID_TOKEN_MAX_AGE)
+        if utc_timestamp > id_token['iat'] + iat_leeway:
+            raise AuthTokenError(self, 'Incorrect id_token: iat')
+
+    def validate_and_return_user_token(self, user_token):
+        client_id, client_secret = self.get_key_and_secret()
+        key = self.find_agent_valid_key(user_token)
+
+        if not key:
+            raise AuthTokenError(self, 'Signature verification failed')
+
+        alg = key['alg']
+        rsakey = jwk.construct(key)
+
+        try:
+            return jwt.decode(
+                user_token,
+                rsakey.to_pem().decode('utf-8'),
+                algorithms=[alg],
+                audience=client_id,
+                issuer=[self.strategy.session_get(self.name + '_agent'),
+                        'https://' + self.strategy.session_get(self.name + '_agent'),
+                        self.strategy.session_get(self.name + '_authority').replace('https://', '')]
+            )
+        except ExpiredSignatureError:
+            raise AuthTokenError(self, 'Signature has expired')
+        except JWTClaimsError as error:
+            raise AuthTokenError(self, str(error))
+        except JWTError:
+            raise
+
+    @handle_http_errors
+    def user_data(self, access_token, *args, **kwargs):
+        user_token = requests.get(self.userinfo_url(), headers={
```

Comment:
This one requires rewrite to utilize distributed claims.
Essential is that initial access token can be only considered valid for authorities' userinfo, but the one to be used with Agent can be different and will be provided only by distributed claims object.
See also: https://gitlab.com/ID4me/general/issues/25